### PR TITLE
Microsoft Teams Webhook alert destination

### PR DESCRIPTION
**Pull request by kevinchiang**
## What type of PR is this? (check all applicable)
- [x] New Alert Destination

## Description
Create a new alert destination for Microsoft Teams Webhooks

## Related Tickets & Documents
https://discuss.redash.io/t/microsoft-teams-integration/852

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
Screen Shot 2022-01-21 at 3 31 15 PM
Screen Shot 2022-01-21 at 3 33 26 PM
Screen Shot 2022-01-21 at 3 33 35 PM

### CODE REVIEW

**Comment by kevinchiang** [comment]
Working on text formatting and icons.

**Comment by kevinchiang** [comment]
microsoftteamswebhook
Planning to use this icon.

**Comment by kevinchiang** [comment]
Screen Shot 2022-01-21 at 4 16 49 PM
Screen Shot 2022-01-21 at 4 16 58 PM
Screen Shot 2022-01-21 at 4 18 27 PM
Screen Shot 2022-01-21 at 4 22 29 PM

**Comment by kevinchiang** [comment]
MS Alert Webhook documentation here: https://docs.microsoft.com/en-us/outlook/actionable-messages/message-card-reference

File: redash/destinations/microsoft_teams_webhook.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    @classmethod
+    def type(cls):
+        return "microsoft_teams_webhook"
+
+    @classmethod
+    def configuration_schema(cls):
+        return {
+            "type": "object",
+            "properties": {
+                "url": {"type": "string"},
```
Comment (arikfr) [inline redash/destinations/microsoft_teams_webhook.py:23]: You might want to add a `title` here to indicate this is a webhook URL.

File: redash/destinations/microsoft_teams_webhook.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    @classmethod
+    def icon(cls):
+        return "fa-bolt"
+
+    def notify(self, alert, query, user, new_state, app, host, options):
+        """
+        :type app: redash.Redash
+        """
+        try:
+            data = {
```
Comment (arikfr) [inline redash/destinations/microsoft_teams_webhook.py:37]: What about supporting custom templates?

**Review by arikfr — COMMENTED** [review]
Thank you for submitting this!

File: redash/destinations/microsoft_teams_webhook.py (hunk 1 above)
Comment (kevinchiang) [reply redash/destinations/microsoft_teams_webhook.py]: Done.

Screen Shot 2022-01-26 at 3 23 28 PM

File: redash/destinations/microsoft_teams_webhook.py (hunk 2 above)
Comment (kevinchiang) [reply redash/destinations/microsoft_teams_webhook.py:37]: I can try to copy what the chatwork destination did for message templates.

**Comment by kevinchiang** [comment]
@arikfr , one thing I noticed during testing is that a trigger is fired when the criteria is met _and_ when the criteria is reset.  Is that correct?  Is that info in the new_state argument?

File: redash/destinations/microsoft_teams_webhook.py (hunk 2 above)
Comment (kevinchiang) [reply redash/destinations/microsoft_teams_webhook.py]: Screen Shot 2022-01-27 at 10 19 36 AM

**Comment by susodapop** [comment]
> I noticed during testing is that a trigger is fired when the criteria is met and when the criteria is reset. Is that correct?

Yes. 

> Is that info in the new_state argument (new_state seems the three options here:

Yes. Related [docs](https://redash.io/help/user-guide/alerts/setting-up-an-alert):

_UNKNOWN means Redash does not have enough data to evaluate the alert criteria. You will see this status immediately after creating your Alert until the query has executed. You will also see this status if there was no data in the query result or if the most recent query result doesn’t include the Value Column you configured._

**Comment by jnmclarty** [comment]
If for some reason this doesn't make it into the next release, could the patch from #5726 ship tactically (it's super tiny), and then reverted in this PR?

**Comment by DougTrajano** [comment]
Any update about this work? There's an official way to implement this MS Teams in the current release (such as a plugin)?

**Comment by susodapop** [comment]
This will be included in the next release later this summer. 

Yes, alert destinations are pluggable just like query runners 👌 This one should be compatible with any installation of Redash V10.

**Comment by DougTrajano** [comment]
> This will be included in the next release later this summer.
> 
> Yes, alert destinations are pluggable just like query runners 👌 This one should be compatible with any installation of Redash V10.

how can I add it to my current instance so? Do we have a tutorial for that?

**Comment by susodapop** [comment]
>  how can I add it to my current instance so? Do we have a tutorial for that?

We don't but working on publishing a doc with a strategy for this. There are a few different approaches. Probably the easiest is to copy the file into your local copy of the source code and then issue a `COPY` command in the Dockerfile that places it into your containers.

**Comment by kevinchiang** [comment]
In case anyone wants to bring this into their existing install, your Dockerfile should look something like this.  (double check the base image)

```
FROM redash/redash:8.0.0.b32245

USER root

# upgrade to fix vulnerabilities and enable ldap
RUN apt update && apt install -y sudo linux-libc-dev
RUN pip install --upgrade pip && pip install --upgrade ldap3 Werkzeug==0.16.1

USER redash

COPY microsoft_teams_webhook.py /app/redash/destinations/microsoft_teams_webhook.py
COPY microsoft_teams_webhook.png /app/client/dist/images/destinations/microsoft_teams_webhook.png
```

Then set the environment variable
```
REDASH_ADDITIONAL_DESTINATIONS: redash.destinations.microsoft_teams_webhook
```

**Comment by arikfr** [comment]
Thanks!

**Pull request merged**